rllib/core/rl_module/marl_module.py

```diff
@@ -0,0 +1,295 @@
+
+
+    Args:
+        config: The config dict for the multi-agent RLModule (see above).
+    """
+
+    def __init__(self, config: Mapping[str, Any], **kwargs) -> None:
+        super().__init__(config, **kwargs)
+
+        self._rl_modules: Mapping[ModuleID, RLModule] = self._make_modules()
+        self._trainable_rl_modules = self.config.get(
+            "trainable_modules", set(self._rl_modules.keys())
+        )
+
+    def keys(self) -> Iterator[ModuleID]:
+        """Returns an iteratable of module ids."""
+        return self._rl_modules.keys()
+
+    def get_trainable_module_ids(self) -> Set[ModuleID]:
+        """Returns the set of ids of the trainable modules."""
+        return self._trainable_rl_modules
+
+    @override(RLModule)
+    def make_distributed(self, dist_config: Mapping[str, Any] = None) -> None:
+        """Makes the module distributed.
+
+        It loops through all the modules and calls their make_distributed method.
+
+        Args:
+            dist_config: The optional distributed configuration to use for all make
+                distributed calls.
+        """
+        for module in self._rl_modules.values():
+            module.make_distributed(dist_config)
+
+    @override(RLModule)
+    def is_distributed(self) -> bool:
+        """Returns True if all sub-modules are distributed."""
+        return all(module.is_distributed() for module in self._rl_modules.values())
+
+    @override(RLModule)
+    def get_state(self) -> Mapping[str, Any]:
+        """Returns the state of the multi-agent module.
+
+        The default implementation loops all modules and calls their get_state method.
+        Override this method if you want to change the get_state behavior.
+
+        Returns:
+            A nested state dict with the first layer being the module ID and the second
+            is the state of the module.
+        """
+        return {
+            module_id: module.get_state()
+            for module_id, module in self._rl_modules.items()
+        }
+
+    @override(RLModule)
+    def set_state(self, state_dict: Mapping[str, Any]) -> None:
+        """Sets the state dict of the multi-agent module.
+
+        The default implementation is a mapping from independent module IDs to their
+        individual state_dicts. Override this method to customize the state_dict for
+        custom more advanced multi-agent use cases.
+
+        Args:
+            state_dict: The state dict to set.
+        """
+        for module_id, module in self._rl_modules.items():
+            module.set_state(state_dict[module_id])
+
+    @override(RLModule)
+    def as_multi_agent(self) -> "MultiAgentRLModule":
+        """Returns a multi-agent wrapper around this module.
+
+        This method is overridden to avoid double wrapping.
+
+        Returns:
+            The instance itself.
+        """
+        return self
+
+    def add_module(
+        self,
+        module_id: ModuleID,
+        module: RLModule,
```

Comment:
Alternatively, pass in a ModuleConfig object?

I'm not 100% sure about this, but we should be consistent: If we allow passing in a config in the c'tor, then we should be allowed to do this here, too.